telethon/tl/session.py

```diff
@@ -54,49 +63,181 @@ def __init__(self, session_user_id):
             self.save_entities = True
             self.flood_sleep_threshold = 60
 
-        # Cross-thread safety
-        self._seq_no_lock = Lock()
-        self._msg_id_lock = Lock()
-        self._save_lock = Lock()
-
         self.id = helpers.generate_random_long(signed=True)
         self._sequence = 0
         self.time_offset = 0
         self._last_msg_id = 0  # Long
 
+        # Cross-thread safety
+        self._seq_no_lock = Lock()
+        self._msg_id_lock = Lock()
+        self._db_lock = Lock()
+
         # These values will be saved
-        self.server_address = None
-        self.port = None
-        self.auth_key = None
-        self.layer = 0
-        self.salt = 0  # Signed long
-        self.entities = EntityDatabase()  # Known and cached entities
+        self._server_address = None
+        self._port = None
+        self._auth_key = None
+        self._layer = 0
+        self._salt = 0  # Signed long
+
+        # Migrating from .json -> SQL
+        entities = self._check_migrate_json()
+
+        self._conn = sqlite3.connect(self.filename, check_same_thread=False)
+        c = self._conn.cursor()
+        c.execute("select name from sqlite_master "
+                  "where type='table' and name='version'")
+        if c.fetchone():
+            # Tables already exist, check for the version
+            c.execute("select version from version")
+            version = c.fetchone()[0]
+            if version != CURRENT_VERSION:
+                self._upgrade_database(old=version)
+                self.save()
+
+            # These values will be saved
+            c.execute('select * from sessions')
+            self._server_address, self._port, key, \
+                self._layer, self._salt = c.fetchone()
+
+            from ..crypto import AuthKey
+            self._auth_key = AuthKey(data=key)
+            c.close()
+        else:
+            # Tables don't exist, create new ones
+            c.execute("create table version (version integer)")
+            c.execute(
+                """create table sessions (
```

Comment:
@delivrance do you think I should save `dc_id` here as a primary key so I could have several `auth_key`'s?

Reply:
Anything globally unique is valid, so `dc_id` as a single primary key sounds like a reasonable choice.

Reply:
Then I need a way to pass the DC ID as well. Which can get a bit strange because they default to `None` and then a single IP is set.